apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -576,29 +608,96 @@ async Task PrepareNewSessionAsync()
                         minimumSizeWaiters.ForEach(tcs => tcs.TrySetException(e));
                     }
                 }
+            } 
+
+            private Task GetNursePoolBackToHealthTask(CancellationToken cancellationToken)
+            {
+                if (Healthy || Shutdown)
+                {
+                    return Task.FromResult(1);
+                }
+                // If already being nursed back to health, wait for that to be done.
+                var newTcs = new TaskCompletionSource<int>();
+                var previousTcs = Interlocked.CompareExchange(ref _nurseBackToHealthTaks, newTcs, null);
+                if (previousTcs != null)
+                {
+                    return previousTcs.WithCancellationToken(cancellationToken);
+                }
+
+                // We want to try and nurse the pool back to health even if the (first) caller cancels the call.
+                // That is if the caller cancels the call, we want to stop waiting for the nursing but we actually
+                // want the nursing to continue.
+                Parent.ConsumeBackgroundTask(Task.Run(async () =>
+                {
+                    try
+                    {
+                        // Not using the cancellation token here so this can go on even if the caller cancels the call.
+                        var sessions = await CreatePooledSessionsAsync(LastUnhealthyBatch, CancellationToken.None).ConfigureAwait(false);
+                        foreach (var session in sessions)
+                        {
+                            ReleaseInactiveSession(session, maybeCreateReadWriteTransaction: true);
+                        }
+
+                        // We have succesfully nurse the pool back to health.
+                        newTcs.TrySetResult(1);
+                    }
+                    // Note: we expect this to always actually be an RpcException, but we don't want to end up unhealthy
+                    // with a lot of 
+                    catch (Exception e)
+                    {
+                        // We couldn't nurse the pool back to health.
+                        newTcs.TrySetException(e);
+
+                        Parent._logger.Warn(() => $"Failed to batch create session for {_databaseName}", e);
+
+                        // If we are here it's because the pool was unhealthy, so we are not receiving and don't have
```

Comment:
I don't understand this comment. I suspect I could after reading more context, but if it can be expanded so I don't have to, that would be great :)

Reply:
Done.